File: examples/cpp/tui_minimal.cpp
// tui_minimal.cpp — the mechanics of a TUI that does not fall apart. ncurses.
//
// This is NOT a design. It is deliberately ugly. It exists so you never have
// to debug the four things that make people give up on curses, and can spend
// your time on the part we are actually grading. The Python twin is
// examples/python/tui_minimal.py — same structure, same comments.
//
//   1. RESIZE. Terminals change size while you are running. ncurses delivers
//      KEY_RESIZE; ignore it and your layout is corrupt for the rest of the
//      session. Drag your terminal edge while this runs.
//
//   2. TOO SMALL. At some point the window cannot hold your layout. Garbling
//      is not acceptable and neither is crashing. Say so, in words, and
//      recover the moment there is room again.
//
//   3. NO COLOR. `NO_COLOR=1 ./tui_minimal` must still be fully usable, and so
//      must a monochrome terminal. Roughly 1 in 12 men has a color vision
//      deficiency, and your operator may be on a washed-out projector. Every
//      state here carries a TEXT marker as well as a color. Never encode
//      meaning in color alone.
//
//   4. INPUT LATENCY. nodelay() makes getch() non-blocking so the render loop
//      never stalls waiting for a keypress. A TUI that pauses its animation
//      while you hold a key feels broken.
//
// Build:  make tui        (needs ncurses: apt install libncurses-dev,
//                          brew install ncurses, or use FTXUI on Windows)
// Keys:   q quit   ? help   space pause   r restart

#include <ncurses.h>

#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <string>
#include <thread>
#include <vector>

#include "scenario.hpp"

using namespace lm;

// Semantic names, not color names. When you decide FAULT should be orange
// instead of red you change one line and nothing else mentions a color. This
// is the smallest possible design-token file — yours will be bigger and
// SHARED WITH ALL THREE SURFACES.
enum Token { TOK_NOMINAL = 1, TOK_DEGRADED, TOK_FAULT, TOK_MUTED };
static const char* MARK[] = {"", "[ OK ]", "[WARN]", "[FAIL]", "      "};

static constexpr int MIN_W = 60, MIN_H = 16;

struct App {
    ScenarioPlayer& player;
    bool use_color = false, paused = false, show_help = false;
    std::deque<ScenarioPlayer::Note> log;
    int h = 0, w = 0;
    double pause_started = 0;

    explicit App(ScenarioPlayer& p) : player(p) {
        use_color = has_colors() && std::getenv("NO_COLOR") == nullptr;
        curs_set(0);
        nodelay(stdscr, TRUE);          // (4) never block the loop on input
        keypad(stdscr, TRUE);
        if (use_color) {
            start_color();
            use_default_colors();
            init_pair(TOK_NOMINAL,  COLOR_GREEN,  -1);
            init_pair(TOK_DEGRADED, COLOR_YELLOW, -1);
            init_pair(TOK_FAULT,    COLOR_RED,    -1);
            init_pair(TOK_MUTED,    COLOR_CYAN,   -1);
        }
        on_resize();
    }

    void on_resize() { getmaxyx(stdscr, h, w); clear(); }   // (1)

    int attr(Token t) const {
        if (!use_color) return (t == TOK_FAULT) ? A_BOLD : A_NORMAL;
        return COLOR_PAIR(t);
    }

    // Clipped write. ncurses errors if you write off-screen, so clip, always.
    void put(int y, int x, const std::string& s, Token t = TOK_MUTED, bool bold = false) {
        if (y < 0 || y >= h || x >= w) return;
        std::string txt = s.substr(0, std::max(0, w - x - 1));
        if (txt.empty()) return;
        int a = attr(t) | (bold ? A_BOLD : 0);
        attron(a); mvaddstr(y, x, txt.c_str()); attroff(a);
    }

    static Token sensor_token(const std::string& v) {
        if (v == "ok") return TOK_NOMINAL;
        if (v == "degraded") return TOK_DEGRADED;
        if (v == "down" || v == "lost") return TOK_FAULT;
        return TOK_MUTED;
    }

    void draw_too_small() {                                  // (2)
        erase();
        char msg[96];
        std::snprintf(msg, sizeof msg, "Window too small: %dx%d, need %dx%d", w, h, MIN_W, MIN_H);
        put(h / 2, std::max(0, (w - int(strlen(msg))) / 2), msg, TOK_DEGRADED, true);
        const char* hint = "resize the terminal or press q";
        put(h / 2 + 1, std::max(0, (w - int(strlen(hint))) / 2), hint);
        refresh();
    }

    void draw(const Tick& tk) {
        erase();
        put(0, 0, std::string(std::max(0, w - 1), '-'));
        put(0, 2, " ROVER-01  " + tk.state + " ", TOK_MUTED, true);
        char clock[64];
        std::snprintf(clock, sizeof clock, " t=%6.1fs %s ", tk.t, paused ? "[PAUSED]" : "");
        put(0, std::max(0, w - int(strlen(clock)) - 2), clock);

        // (3) every state gets a TEXT marker as well as a color
        double b = tk.battery_pct;
        Token bt = b < 8 ? TOK_FAULT : (b < 20 ? TOK_DEGRADED : TOK_NOMINAL);
        char line[128];
        std::snprintf(line, sizeof line, "BATTERY %s %5.1f%%", MARK[bt], b);
        put(2, 2, line, bt, bt == TOK_FAULT);
        int bar = std::clamp(w - 34, 10, 30);
        int filled = int(bar * b / 100.0 + 0.5);
        put(2, 30, std::string(filled, '#') + std::string(bar - filled, '.'), bt);

        int y = 4;
        const char* names[] = {"lidar", "imu", "comms"};
        const std::string vals[] = {tk.lidar, tk.imu, tk.comms};
        for (int i = 0; i < 3; ++i) {
            Token t = sensor_token(vals[i]);
            std::snprintf(line, sizeof line, "%-6s %s %s", names[i], MARK[t], vals[i].c_str());
            put(y++, 2, line, t);
        }

        ++y;
        put(y, 2, "TASK    " + (tk.task.empty() ? std::string("(none)") : tk.task));
        if (tk.task_total) {
            std::snprintf(line, sizeof line, "        %d/%d", tk.task_idx, tk.task_total);
            put(y + 1, 2, line);
        }
        std::snprintf(line, sizeof line, "POSE    x=%6.2f  y=%6.2f  hdg=%5.1f  v=%.2f",
                      tk.x, tk.y, tk.heading, tk.speed);
        put(y + 2, 2, line);

        if (tk.estop) put(y + 4, 2, "  *** EMERGENCY STOP ENGAGED ***  ", TOK_FAULT, true);

        int ly = y + 6;
        put(ly, 2, "EVENTS", TOK_MUTED, true);
        int room = std::max(1, h - ly - 3);
        int start = std::max(0, int(log.size()) - room);
        for (int i = start; i < int(log.size()); ++i) {
            std::snprintf(line, sizeof line, "%6.1f  %s", log[i].first, log[i].second.c_str());
            put(ly + 1 + (i - start), 2, line);
        }

        put(h - 1, 2, "q quit   ? help   space pause   r restart");
        if (show_help) draw_help();
        refresh();
    }

    void draw_help() {
        const std::vector<std::string> lines = {
            "KEYS", "", "  q      quit", "  ?      toggle this help",
            "  space  pause / resume", "  r      restart scenario", "",
            "  Esc or ? closes this overlay."};
        size_t bw = 0;
        for (auto& s : lines) bw = std::max(bw, s.size());
        bw += 4;
        int bh = int(lines.size()) + 2;
        int y0 = std::max(0, (h - bh) / 2), x0 = std::max(0, (w - int(bw)) / 2);
        for (int i = 0; i < bh; ++i) put(y0 + i, x0, std::string(bw, ' '));
        for (size_t i = 0; i < lines.size(); ++i)
            put(y0 + 1 + int(i), x0 + 2, lines[i], TOK_MUTED, i == 0);
    }

    void run() {
        while (true) {
            double now = ScenarioPlayer::now();
            int ch;
            while ((ch = getch()) != ERR) {
                if (ch == KEY_RESIZE) on_resize();
                else if (ch == 'q' || ch == 'Q') return;
                else if (ch == '?') show_help = !show_help;
                else if (ch == 27) show_help = false;       // Esc always backs out
                else if (ch == ' ') {
                    paused = !paused;
                    if (paused) pause_started = ScenarioPlayer::now();
                    else player.nudge_clock(ScenarioPlayer::now() - pause_started);
                } else if (ch == 'r' || ch == 'R') { player.restart(); log.clear(); }
            }

            if (w < MIN_W || h < MIN_H) {
                draw_too_small();
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
                continue;
            }

            static Tick frozen;
            const Tick* tk;
            if (!paused) {
                tk = &player.poll();
                frozen = *tk;
                for (auto& n : player.drain_notes()) {
                    log.push_back(n);
                    if (log.size() > 200) log.pop_front();
                }
            } else {
                tk = &frozen;
            }
            draw(*tk);

            double spent = ScenarioPlayer::now() - now;
            if (spent < 1.0 / 30)
                std::this_thread::sleep_for(std::chrono::duration<double>(1.0 / 30 - spent));
        }
    }
};

int main(int argc, char** argv) {
    const char* path = "../../scenario/scenario.jsonl";
    double speed = 1.0;
    for (int i = 1; i < argc; ++i) {
        if (!strcmp(argv[i], "--file") && i + 1 < argc) path = argv[++i];
        else if (!strcmp(argv[i], "--speed") && i + 1 < argc) speed = atof(argv[++i]);
    }
    ScenarioPlayer player(path, speed, /*loop=*/true);
    initscr();
    noecho();
    cbreak();
    App app(player);
    app.run();
    endwin();
    return 0;
}

File: examples/python/tui_minimal.py
"""
tui_minimal.py — the mechanics of a TUI that does not fall apart.

This is NOT a design. It is deliberately ugly. It exists so you never have to
debug the four things that make people give up on curses, and can spend your
time on the part we are actually grading.

The four things:

  1. RESIZE. Terminals change size while you are running. curses delivers
     KEY_RESIZE; if you ignore it your layout is corrupt for the rest of the
     session. Handled in `_on_resize`. Drag your terminal edge while this runs.

  2. TOO SMALL. At some point the window cannot hold your layout. Garbling is
     not acceptable and neither is crashing. Say so, in words, and recover the
     moment there is room again. See `_draw_too_small`.

  3. NO COLOR. `NO_COLOR=1 python tui_minimal.py` must still be fully usable,
     and so must a monochrome terminal. This is not just accessibility
     box-ticking: roughly 1 in 12 men has a color vision deficiency, and your
     operator may be on a washed-out projector. Every state here carries a
     TEXT marker as well as a color. Never encode meaning in color alone.

  4. INPUT LATENCY. `nodelay(True)` makes getch() non-blocking so the render
     loop never stalls waiting for a keypress. A TUI that pauses its animation
     while you hold a key feels broken.

Run:
    python tui_minimal.py
    NO_COLOR=1 python tui_minimal.py
    python tui_minimal.py --speed 4

Keys: q quit,  ? help,  space pause,  r restart

Windows note: the stdlib `curses` module is not bundled with Python on
Windows. Either run this under WSL, or `pip install windows-curses`, or write
your TUI with a cross-platform library. Python options that need no curses:
`rich` / `textual`. C++ options: ncursesw, or FTXUI (header-only, builds on
Windows). Any of these is fine for your submission.
"""

import argparse
import curses
import os
import time

from scenario_player import ScenarioPlayer

# Semantic names, not color names. When you decide FAULT should be orange
# instead of red you change one line, and nothing else in the file mentions
# a color. This is the smallest possible version of a design-token file --
# yours will be bigger and SHARED WITH ALL THREE SURFACES.
TOKENS = {
    "nominal":  {"pair": 1, "fg": curses.COLOR_GREEN,  "mark": "[ OK ]"},
    "degraded": {"pair": 2, "fg": curses.COLOR_YELLOW, "mark": "[WARN]"},
    "fault":    {"pair": 3, "fg": curses.COLOR_RED,    "mark": "[FAIL]"},
    "muted":    {"pair": 4, "fg": curses.COLOR_CYAN,   "mark": "      "},
}

MIN_W, MIN_H = 60, 16


class App:
    def __init__(self, stdscr, player):
        self.scr = stdscr
        self.player = player
        self.paused = False
        self.show_help = False
        self.log = []
        self.use_color = curses.has_colors() and not os.environ.get("NO_COLOR")

        curses.curs_set(0)
        stdscr.nodelay(True)          # (4) never block the loop on input
        stdscr.keypad(True)
        if self.use_color:
            curses.start_color()
            curses.use_default_colors()
            for tok in TOKENS.values():
                curses.init_pair(tok["pair"], tok["fg"], -1)
        self._on_resize()

    # -- (1) resize --------------------------------------------------------

    def _on_resize(self):
        self.h, self.w = self.scr.getmaxyx()
        self.scr.clear()

    def attr(self, token):
        if not self.use_color:
            return curses.A_BOLD if token == "fault" else curses.A_NORMAL
        return curses.color_pair(TOKENS[token]["pair"])

    def put(self, y, x, text, token="muted", bold=False):
        """Clipped write. curses raises if you touch the last cell of the last
        line, and it raises if you write off-screen at all -- so clip, always."""
        if y < 0 or y >= self.h:
            return
        text = text[: max(0, self.w - x - 1)]
        if not text:
            return
        a = self.attr(token) | (curses.A_BOLD if bold else 0)
        try:
            self.scr.addstr(y, x, text, a)
        except curses.error:
            pass

    # -- rendering ---------------------------------------------------------

    def _draw_too_small(self):
        """(2) Honest degradation. Not a crash, not a garbled screen."""
        self.scr.erase()
        msg = f"Window too small: {self.w}x{self.h}, need {MIN_W}x{MIN_H}"
        self.put(max(0, self.h // 2), max(0, (self.w - len(msg)) // 2), msg, "degraded", bold=True)
        hint = "resize the terminal or press q"
        self.put(max(0, self.h // 2) + 1, max(0, (self.w - len(hint)) // 2), hint, "muted")
        self.scr.refresh()

    def sensor_token(self, value):
        return {"ok": "nominal", "degraded": "degraded", "down": "fault", "lost": "fault"}.get(value, "muted")

    def draw(self, tick, dt):
        self.scr.erase()

        # header
        self.put(0, 0, "─" * (self.w - 1), "muted")
        self.put(0, 2, f" ROVER-01  {tick['state']} ", "muted", bold=True)
        clock = f" t={tick['t']:6.1f}s {'[PAUSED]' if self.paused else ''} "
        self.put(0, max(0, self.w - len(clock) - 2), clock, "muted")

        # (3) every state gets a TEXT marker as well as a color
        batt = tick["battery_pct"]
        btok = "fault" if batt < 8 else ("degraded" if batt < 20 else "nominal")
        bar_w = max(10, min(30, self.w - 34))
        filled = int(round(bar_w * batt / 100))
        self.put(2, 2, f"BATTERY {TOKENS[btok]['mark']} {batt:5.1f}%", btok, bold=(btok == "fault"))
        self.put(2, 30, "█" * filled + "░" * (bar_w - filled), btok)

        y = 4
        for name in ("lidar", "imu", "comms"):
            v = tick[name]
            tok = self.sensor_token(v)
            self.put(y, 2, f"{name.upper():<6} {TOKENS[tok]['mark']} {v}", tok)
            y += 1

        y += 1
        self.put(y, 2, f"TASK    {tick['task'] or '(none)'}", "muted")
        if tick["task_total"]:
            self.put(y + 1, 2, f"        {tick['task_idx']}/{tick['task_total']}", "muted")
        self.put(y + 2, 2, f"POSE    x={tick['x']:6.2f}  y={tick['y']:6.2f}  "
                           f"hdg={tick['heading']:5.1f}  v={tick['speed']:.2f}", "muted")

        if tick["estop"]:
            self.put(y + 4, 2, "  *** EMERGENCY STOP ENGAGED ***  ", "fault", bold=True)

        # event log -- the discrete half of the data
        ly = y + 6
        self.put(ly, 2, "EVENTS", "muted", bold=True)
        for i, (t, note) in enumerate(self.log[-max(1, self.h - ly - 3):]):
            self.put(ly + 1 + i, 2, f"{t:>6.1f}  {note}", "muted")

        self.put(self.h - 1, 2, "q quit   ? help   space pause   r restart", "muted")

        if self.show_help:
            self._draw_help()
        self.scr.refresh()

    def _draw_help(self):
        lines = ["KEYS", "", "  q      quit", "  ?      toggle this help",
                 "  space  pause / resume", "  r      restart scenario", "",
                 "  Esc or ? closes this overlay."]
        bw = max(len(s) for s in lines) + 4
        bh = len(lines) + 2
        y0, x0 = max(0, (self.h - bh) // 2), max(0, (self.w - bw) // 2)
        for i in range(bh):
            self.put(y0 + i, x0, " " * bw, "muted")
        for i, s in enumerate(lines):
            self.put(y0 + 1 + i, x0 + 2, s, "muted", bold=(i == 0))

    # -- loop --------------------------------------------------------------

    def run(self):
        last = time.monotonic()
        pause_started = None
        while True:
            now = time.monotonic()
            dt = now - last
            last = now

            ch = self.scr.getch()
            while ch != -1:
                if ch == curses.KEY_RESIZE:
                    self._on_resize()
                elif ch in (ord("q"), ord("Q")):
                    return
                elif ch == ord("?"):
                    self.show_help = not self.show_help
                elif ch == 27:                      # Esc always backs out
                    self.show_help = False
                elif ch == ord(" "):
                    self.paused = not self.paused
                    if self.paused:
                        pause_started = time.monotonic()
                    elif pause_started:
                        self.player._t0 += time.monotonic() - pause_started
                elif ch in (ord("r"), ord("R")):
                    self.player.restart()
                    self.log.clear()
                ch = self.scr.getch()

            if self.w < MIN_W or self.h < MIN_H:
                self._draw_too_small()
                time.sleep(1 / 20)
                continue

            if not self.paused:
                tick = self.player.poll()
                self.log.extend(self.player.drain_notes())
            else:
                tick = self.player.ticks[self.player._idx]

            self.draw(tick, dt)
            time.sleep(max(0.0, 1 / 30 - (time.monotonic() - now)))


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--file", default="../../scenario/scenario.jsonl")
    ap.add_argument("--speed", type=float, default=1.0)
    args = ap.parse_args()
    player = ScenarioPlayer(args.file, speed=args.speed, loop=True)
    curses.wrapper(lambda scr: App(scr, player).run())


if __name__ == "__main__":
    main()
